File: tests/pointer.js
module.exports = [
  `i32 swap(i32 *c, i32 *d) {
    i32 tmp = *c;
    *c = *d;
    *d = tmp;
    return (0);
  };
  export i32 main() {
    i32 a = 10;
    i32 b = 20;
    swap(&a, &b);
    return (b);
  };`,
  `10`,
  `i32 swap(i32 *c, i32 *d) {
    i32 tmp = *c;
    *c = *d;
    *d = tmp;
    return (0);
  };
  i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrB = &b;
    i32 res = swap(&a, ptrB);
    return (*ptrB);
  };
  `,
  `42`,
  `i32 swap(i32 *c, i32 *d) {
    i32 tmp = *c;
    *c = *d;
    *d = tmp;
    return (0);
  };
  i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    i32 res = swap(ptrA, ptrB);
    return (*ptrA);
  };`,
  `66`,
  `i32 swap(i32 *c, i32 *d) {
    i32 tmp = *c;
    *c = *d;
    *d = tmp;
    return (0);
  };
  i32 main() {
    i32 a = 42;
    i32 c = 55;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrC = &c;
    i32 *ptrB = &b;
    i32 res = swap(ptrA, &b);
    return (*ptrA);
  };`,
  `66`,
  `export i32 main() {
    i32 a = 10;
    i32 b = 20;
    return (&a);
  };`,
  `0`,
  `export i32 main() {
    i32 a = 10;
    i32 b = 20;
    return (&b);
  };`,
  `4`,
  `export i32 main() {
    i32 a = 10;
    i32 b = 20;
    i32 *ptr = &a;
    return (ptr);
  };`,
  `0`,
  `export i32 main() {
    i32 a = 10;
    i32 b = 20;
    i32 *ptr = &b;
    return (ptr);
  };`,
  `4`,
  `export i32 main() {
    i32 a = 10;
    i32 b = 20;
    i32 *ptr = &b;
    return (*ptr);
  };`,
  `20`,
  ``,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    ptrA = ptrB;
    return (*ptrA);
  };
  `,
  `66`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    ptrA = ptrB;
    return (*ptrA);
  };
  `,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    ptrA = ptrB;
    return (ptrA == ptrB);
  };
  `,
  `1`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    ptrA = &b;
    return (ptrA == ptrB);
  };
  `,
  `1`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    ptrA = ptrB;
    return (*ptrA);
  };
  `,
  `66`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &a;
    ptrB = &b;
    return (*ptrB);
  };`,
  `66`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrC = &a;
    i32 *ptrA = &b;
    i32 *ptrB = &b;
    ptrA = ptrC;
    return (*ptrA);
  };`,
  `42`,
  `i32 main() {
    i32 a = 42; // 0x0
    i32 b = 66; // 0x4
    i32 *ptrA = 0x0;
    i32 *ptrB = 0x4;
    return (*ptrB);
  };`,
  `66`,
  `i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptrA = &a;
    i32 *ptrB = &b;
    *ptrB = 77;
    return (*ptrB + b);
  };`,
  `154`,
  `i32 swap(i32 *c, i32 *d) {
    i32 *e = c;
    *(*(&e)) = *d;
    return (0);
  };
  i32 main() {
    i32 a = 42;
    i32 b = 66;
    i32 *ptr = &b;
    i32 res = swap(&a, ptr);
    return ((*(&a)) + b);
  };`,
  `132`,
  `i32 main() {
    i32 num = 10;
    i32 *ptr1 = &num;
    i32 *ptr2 = &ptr1;
    return (**ptr2 == *ptr1);
  };`,
  `1`,
  `i32 main() {
    i32 x = 0;
    i32 y = 0;
    i32 *p = &x;
    i32 *q = &y;
    i32 *pp = &p;
    pp = &q;
    **pp = 4;
    return (y == 4 && x == 0);
  };
  `,
  `1`,
  `int main() {
    int a = 100;
    int *ptr = &a;
    return ((&ptr != &a) && ptr == &a);
  };
  `,
  `1`,
  `i32 main() {
    i32 b = 66;
    i32 *ptr = &b;
    return ((&*ptr) == &b);
  };`,
  `1`,
  `i32 main() {
    i32 a = 66;
    i32 b = 77;
    i32 *ptr = &b;
    return (&*ptr == &b);
  };`,
  `1`,
  `i32 main() {
    i32 x = 7;
    i32 a = 66;
    i32 *ptr = &a;
    return (*&ptr == &*ptr);
  };`,
  `1`
];

File: test/pointer.c
int test1a(int *c, int *d) {
  int tmp = *c;
  *c = *d;
  *d = tmp;
  return (0);
};
int test1b() {
  int a = 10;
  int b = 20;
  test1a(&a, &b);
  return (b);
};

int test2a(int *c, int *d) {
  int tmp = *c;
  *c = *d;
  *d = tmp;
  return (0);
};
int test2b() {
  int a = 42;
  int b = 66;
  int *ptrB = &b;
  int res = test2a(&a, ptrB);
  return (*ptrB);
};

int test3a(int *c, int *d) {
  int tmp = *c;
  *c = *d;
  *d = tmp;
  return (0);
};
int test3b() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  int res = test3a(ptrA, ptrB);
  return (*ptrA);
};

int test4a(int *c, int *d) {
  int tmp = *c;
  *c = *d;
  *d = tmp;
  return (0);
};
int test4b() {
  int a = 42;
  int c = 55;
  int b = 66;
  int *ptrA = &a;
  int *ptrC = &c;
  int *ptrB = &b;
  int res = test4a(ptrA, &b);
  return (*ptrA);
};

int test9() {
  int a = 10;
  int b = 20;
  int *ptr = &b;
  return (*ptr);
};

int test10() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  ptrA = ptrB;
  return (*ptrA);
};

int test11() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  ptrA = ptrB;
  return (*ptrA);
};

int test12() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  ptrA = ptrB;
  return (ptrA == ptrB);
};

int test13() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  ptrA = &b;
  return (ptrA == ptrB);
};

int test14() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  ptrA = ptrB;
  return (*ptrA);
};

int test15() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &a;
  ptrB = &b;
  return (*ptrB);
};

int test16() {
  int a = 42;
  int b = 66;
  int *ptrC = &a;
  int *ptrA = &b;
  int *ptrB = &b;
  ptrA = ptrC;
  return (*ptrA);
};

int test17() {
  int a = 42;
  int b = 66;
  int *ptrA = &a;
  int *ptrB = &b;
  *ptrB = 77;
  return (*ptrB + b);
};

int test18a(int *c, int *d) {
  int *e = c;
  *(*(&e)) = *d;
  return (0);
};
int test18b() {
  int a = 42;
  int b = 66;
  int *ptr = &b;
  int res = test18a(&a, ptr);
  return ((*(&a)) + b);
};

int test19() {
  int num = 10;
  int *ptr1 = &num;
  int *ptr2 = ptr1;
  return (*ptr2 == *ptr1);
};

int test20() {
  int x = 0;
  int y = 0;
  int *p = &x;
  int *q = &y;
  int *pp = &p;
  pp = &q;
  **pp = 4;
  return (y == 4 && x == 0);
};

int test21() {
  int a = 100;
  int *ptr = &a;
  return ((&ptr != &a) && ptr == &a);
};

int test22() {
  int b = 66;
  int *ptr = &b;
  return ((&*ptr) == &b);
};

int test23() {
  int a = 66;
  int b = 77;
  int *ptr = &b;
  return (&*ptr == &b);
};

int test24() {
  int x = 7;
  int a = 66;
  int *ptr = &a;
  return (*&ptr == &*ptr);
};

int main() {
  assert(test1b(), 10);
  assert(test2b(), 42);
  assert(test3b(), 66);
  assert(test4b(), 66);
  assert(test9(), 20);
  assert(test10(), 66);
  assert(test11(), 66);
  assert(test12(), 1);
  assert(test13(), 1);
  assert(test14(), 66);
  assert(test15(), 66);
  assert(test16(), 42);
  assert(test17(), 154);
  assert(test18b(), 132);
  assert(test19(), 1);
  assert(test20(), 1);
  assert(test21(), 1);
  assert(test22(), 1);
  assert(test23(), 1);
  assert(test24(), 1);
};
